{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Script to run SISTR: \n",
    "# Method to obtain posterior estimate of s and corresponding p value for each STR\n",
    "\n",
    "### Imports ###\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy.stats import geom\n",
    "import copy\n",
    "import sys\n",
    "sys.path.append(\"/storage/BonnieH/selection_project/helper_functions\")\n",
    "from LRT_functions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Main function ###\n",
    "def main():\n",
    "    # Load parameters\n",
    "    constant_het = 0.005\n",
    "    denom_het = 3\n",
    "    constant_common = 1\n",
    "    denom_common = 1\n",
    "    eps_bins = 0.3\n",
    "    inFile = '/storage/BonnieH/selection_project/ssc_files/0810/DMPK.txt'\n",
    "    use_het = 'y'\n",
    "    use_common = 'n'\n",
    "    use_bins = 'y'\n",
    "    num_bins = 5\n",
    "    abc_model = 'eurodem_prior2' \n",
    "    lrt_model = 'eurodem_0810'\n",
    "    LRT_num_sims = 2000\n",
    "    \n",
    "    \n",
    "    allele_freqs_file = open(inFile, 'r')\n",
    "   \n",
    "    print(\"chrom\" + \"\\t\" + \"start\" + \"\\t\" + \"end\" + \"\\t\" + \"total\" + \"\\t\" + \"period\" + \\\n",
    "                  \"\\t\" + \"optimal_ru\" + \"\\t\" + \"motif\" + \"\\t\" + \"coding\" + \"\\t\"  + 'intron' + \\\n",
    "                  '\\t' + 'UTR5' + '\\t' + 'UTR3' + '\\t' + 'promoter5kb' + '\\t' + 'intergenic' + \\\n",
    "                  \"\\t\" + \"tss_gene\" + \"\\t\" + \"het\" + \"\\t\" + \"common\" + \"\\t\" + \"bins\" + \"\\t\" + \\\n",
    "                  \"ABC_s_median\" + \"\\t\" + \"ABC_s_95%_CI\" + \"\\t\" + \"Num_s_accepted_(max_10000)\" + \"\\t\" + \\\n",
    "                  \"Likelihood_0\" + \"\\t\" + \"Likelihood_s\" + \"\\t\" + \"LR\" + \"\\t\" + \"LogLR\" + \\\n",
    "                  \"\\t\" + \"LRT_p_value\" + \"\\n\")\n",
    "\n",
    "    total_lines_acc = 0 # Total lines with >10 s accepted\n",
    "    total_lines = 0 \n",
    "    lower_0 = 0 # Total lines where lower bound is 0\n",
    "    s_acc = [] # List of percent of s accepted\n",
    "     \n",
    "    # Preprocess ABC lookup table\n",
    "    # Get ABC tables\n",
    "    ABC_tables = {}\n",
    "    #opt_allele_dic = {}\n",
    "    #opt_allele_dic[3] = [5,6,7,8,9,10,11,12,13]\n",
    "    #opt_allele_dic[2] = [11,12,13,14,15,16,17,18,19,20]\n",
    "    #opt_allele_dic[4] = [7,8,9,10]\n",
    "    \n",
    "    opt_allele_dic_w_per = {}\n",
    "    opt_allele_dic_w_per[1] = [(1,12),(1,16),(1,20),(1,24),(1,28),(1,32),(1,36),(1,40),(1,44)]\n",
    "    opt_allele_dic_w_per[3] = [(3,5),(3,6),(3,7),(3,8),(3,9),(3,10),(3,11),(3,12),(3,13)]\n",
    "    opt_allele_dic_w_per[2] = [(2,11),(2,12),(2,13),(2,14),(2,15),(2,16),(2,17),(2,18),(2,19),(2,20)]\n",
    "    opt_allele_dic_w_per[4] = [(4,7),(4,8),(4,9),(4,10)]\n",
    "    \n",
    "    pers = [1,2,3,4]\n",
    "    for per in pers:\n",
    "        for per_opt in opt_allele_dic_w_per[per]:\n",
    "            period = per_opt[0]\n",
    "            opt_allele = per_opt[1]\n",
    "            file = '/gymreklab-tscc/bonnieh/abc/results/' + abc_model + '/' + str(period) + '_' + str(opt_allele) + '.txt' \n",
    "            table = GetABCList(file, num_bins)\n",
    "\n",
    "            ABC_tables[per_opt] = table\n",
    "        \n",
    "        \n",
    "    # Perform ABC on each locus\n",
    "    for line in allele_freqs_file:\n",
    "        # Get information from line\n",
    "        total_lines = total_lines + 1\n",
    "        info = line.strip().split('\\t')\n",
    "        chrom = info[0]\n",
    "        start = int(info[1])\n",
    "        end = int(info[2])\n",
    "        freqs = info[3]\n",
    "        total = info[4]\n",
    "        per = int(info[5])\n",
    "        motif = info[6]\n",
    "        coding = info[7]\n",
    "        intron = info[8]\n",
    "        UTR5 = info[9]\n",
    "        UTR3 = info[10]\n",
    "        promoter5kb = info[11]\n",
    "        intergenic = info[12]\n",
    "        gene = info[13]\n",
    "\n",
    "        list_num = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.3483709273182957, 0.00046992481203007516, 0.0010964912280701754, 0.006422305764411027, 0.0015664160401002505, 0.013157894736842105, 0.12844611528822056, 0.14364035087719298, 0.1787280701754386, 0.06531954887218046, 0.009241854636591478, 0.008771929824561403, 0.005325814536340852, 0.0012531328320802004, 0.0043859649122807015, 0.01331453634085213, 0.024592731829573933, 0.01237468671679198, 0.007832080200501253, 0.006735588972431077, 0.004855889724310777, 0.0045426065162907266, 0.002819548872180451, 0.0021929824561403508, 0.002349624060150376, 0.0009398496240601503, 0.00046992481203007516, 0.0006265664160401002, 0, 0, 0, 0, 0.00015664160401002505]\n",
    "        count_num = 0\n",
    "        for elem in list_num:\n",
    "            count_num = count_num + elem\n",
    "        print(count_num)\n",
    "        list_num_copy = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.3483709273182957, 0.00046992481203007516, 0.0010964912280701754, 0.006422305764411027, 0.0015664160401002505, 0.013157894736842105, 0.12844611528822056, 0.14364035087719298, 0.1787280701754386, 0.06531954887218046, 0.009241854636591478, 0.008771929824561403, 0.005325814536340852, 0.0012531328320802004, 0.0043859649122807015, 0.01331453634085213, 0.024592731829573933, 0.01237468671679198, 0.007832080200501253, 0.006735588972431077, 0.004855889724310777, 0.0045426065162907266, 0.002819548872180451, 0.0021929824561403508, 0.002349624060150376, 0.0009398496240601503, 0.00046992481203007516, 0.0006265664160401002, 0, 0, 0, 0, 0.00015664160401002505]\n",
    "        if list_num != list_num_copy:\n",
    "            print('False')\n",
    "        else: \n",
    "            print('True')\n",
    "        opt_allele_dic = {}\n",
    "        \n",
    "        opt_allele_dic[5] = int(0.3483709273182957*10000)\n",
    "        opt_allele_dic[6] = int(0.00046992481203007516*10000)\n",
    "        opt_allele_dic[7] = int(0.0010964912280701754*10000)\n",
    "        opt_allele_dic[8] = int(0.006422305764411027*10000)\n",
    "        opt_allele_dic[9] = int(0.0015664160401002505*10000)\n",
    "        opt_allele_dic[10] = int(0.013157894736842105*10000)\n",
    "        opt_allele_dic[11] = int(0.12844611528822056*10000)\n",
    "        opt_allele_dic[12] = int(0.14364035087719298*10000)\n",
    "        opt_allele_dic[13] = int(0.1787280701754386*10000)\n",
    "        opt_allele_list = range(5, 14, 1)\n",
    "        for opt_allele in opt_allele_list:\n",
    "            \n",
    "        # Get optimal allele and allele_freqs\n",
    "        \n",
    "            opt_allele, allele_freqs = Process_Freqs_Given_Opt(freqs, per, end, start, opt_allele)\n",
    "            print('Opt allele' + str(opt_allele))\n",
    "            print(allele_freqs)\n",
    "            #numpy.random.choice(numpy.arange(1, 7), p=[0.1, 0.05, 0.05, 0.2, 0.4, 0.2])\n",
    "            freq_string = ','.join(str(round(item, 5)) for item in allele_freqs)\n",
    "        \n",
    "            # Add 0s to allele frequency list if number of alleles less than number of bins\n",
    "            if len(allele_freqs) < num_bins:\n",
    "                num_zeros_to_add = int((num_bins - len(allele_freqs))/2)\n",
    "                for i in range(0, num_zeros_to_add):\n",
    "                    freq_string = '0.0,' + freq_string\n",
    "                    freq_string = freq_string + ',0.0'\n",
    "                    \n",
    "            # Get summary stats\n",
    "            obs_het, obs_common, obs_bins = GetSummStats(freq_string, num_bins)\n",
    "        \n",
    "            print(chrom + '\\t' + str(start) + '\\t' + str(end) + '\\t' + total + '\\t' + str(per) + \\\n",
    "                      '\\t' + str(opt_allele) + '\\t' + motif + '\\t' + coding + '\\t'  + intron + \\\n",
    "                      '\\t' + UTR5 + '\\t' + UTR3 + '\\t' + promoter5kb + '\\t' + intergenic + \\\n",
    "                      '\\t' + gene + '\\t' + str(round(obs_het, 7)) + '\\t' + str(obs_common) + '\\t' + \\\n",
    "                      ','.join(str(round(item,4)) for item in obs_bins) + '\\t')\n",
    "        \n",
    "            # All optimal alleles > 13 have the same mutation rate as optimal allele 13\n",
    "            if per == 3 and opt_allele > 13: \n",
    "                opt_allele = 13\n",
    "            if per == 3 and opt_allele < 5:\n",
    "                opt_allele = 5\n",
    "            \n",
    "        \n",
    "            # Read abcFile line by line and place in lookup table in the form of a list\n",
    "        \n",
    "            abc_list = ABC_tables[(per, opt_allele)]\n",
    "            #abc_list = abc_list[0:opt_allele_dic[opt_allele]]\n",
    "            # Perform ABC\n",
    "            s_ABC, lower_bound, upper_bound, num_accepted, s_accepted = Get_S_ABC(abc_list, \n",
    "                                           obs_het, obs_common, obs_bins, constant_het, \n",
    "                                           denom_het, constant_common, denom_common, eps_bins, use_het, \n",
    "                                           use_common, use_bins)\n",
    "            print('s_accepted')\n",
    "            print(s_accepted)\n",
    "            s_acc.extend(s_accepted)\n",
    "        print('Final estimate of s')\n",
    "        print(s_acc)\n",
    "        '''\n",
    "        # Write N/A if <10 s accepted during ABC\n",
    "        if s_ABC == -1:\n",
    "            s_ABC = 'N/A'\n",
    "            ABC_conf_int = 'N/A'\n",
    "            num_accepted = '<10'\n",
    "            likelihood_0 = 'N/A'\n",
    "            likelihood_s_ABC = 'N/A'\n",
    "            LR = 'N/A'\n",
    "            LogLR = 'N/A'\n",
    "            pval = 'N/A'\n",
    "        \n",
    "            print(str(s_ABC) + '\\t' + ABC_conf_int + '\\t' + str(num_accepted) + '\\t')\n",
    "            print(str(likelihood_0) + '\\t' + str(likelihood_s_ABC) + '\\t' + \\\n",
    "                          str(LR) + '\\t' + str(LogLR) + '\\t' + str(pval) + '\\n')\n",
    "         \n",
    "        # Perform LRT if can get posterior estimate of s\n",
    "        else:\n",
    "            # Round s values < 10^-5 to 0\n",
    "       \n",
    "            if s_ABC < 10**-5:\n",
    "                s_ABC = 0\n",
    "            else:\n",
    "                s_ABC = round(s_ABC, 5)\n",
    "\n",
    "            if lower_bound < 10**-5:\n",
    "                lower_bound = 0\n",
    "            else:\n",
    "                lower_bound = round(lower_bound, 5)\n",
    "\n",
    "            if upper_bound < 10**-5:\n",
    "                upper_bound = 0\n",
    "            else:\n",
    "                upper_bound = round(upper_bound, 5)\n",
    "\n",
    "            if lower_bound == 0:\n",
    "                lower_0 = lower_0 + 1\n",
    "\n",
    "            if lower_bound != -1:\n",
    "                total_lines_acc = total_lines_acc + 1\n",
    "\n",
    "            if num_accepted >= 10:\n",
    "                s_acc.append(num_accepted)\n",
    "\n",
    "            ABC_conf_int = '(' + str(lower_bound) + ' , ' + str(upper_bound) + ')'\n",
    "\n",
    "            print(str(s_ABC) + '\\t' + ABC_conf_int + '\\t' + str(num_accepted) + '\\t')\n",
    "\n",
    "            s_ABC_round = get_LRT_bin(s_ABC)\n",
    "\n",
    "            # Perform LRT\n",
    "\n",
    "            ### Get list of s values in LRT simulations file ###\n",
    "            s_list_available = []\n",
    "            lrtFile = '/gymreklab-tscc/bonnieh/lrt/results/' + lrt_model + '/' \\\n",
    "            + str(per) + '_' + str(opt_allele) + '_freqs.txt' \n",
    "            \n",
    "            lrt_file = open(lrtFile, 'r')\n",
    "            header = lrt_file.readline().strip()\n",
    "\n",
    "            for line in lrt_file:\n",
    "                info = line.strip().split('\\t')\n",
    "                s = float(info[0])\n",
    "                if s not in s_list_available:\n",
    "                    s_list_available.append(s)\n",
    "\n",
    "            # Get nearest s in LRT file to perform LRT\n",
    "            if s_ABC_round not in s_list_available:\n",
    "                s_ABC_round = getNearestS(s_ABC_round, s_list_available)\n",
    "\n",
    "            # Get LRT summary statistic tables for s = 0\n",
    "            \n",
    "            lrtFile_for_s_0 = '/gymreklab-tscc/bonnieh/lrt/results/' + lrt_model + '/' + str(per) + '_' + str(opt_allele) + '_15_freqs.txt' \n",
    "            freqs_list_raw_0 = GetLRTListByRow(lrtFile_for_s_0, 0)\n",
    "            LRT_table_0_het = []\n",
    "            LRT_table_0_common = []\n",
    "            LRT_table_0_bins = []\n",
    "\n",
    "            # Get summary statistics from allele frequencies\n",
    "            for freq_string in freqs_list_raw_0:\n",
    "\n",
    "                obs_het_0, obs_common_0, obs_bins_0 = GetSummStats(freq_string, num_bins)\n",
    "                LRT_table_0_het.append(obs_het_0) \n",
    "                LRT_table_0_common.append(obs_common_0) \n",
    "                LRT_table_0_bins.append(obs_bins_0)\n",
    "\n",
    "            # Get LRT summary statistic tables for s = s_ABC_round\n",
    "            if s_ABC_round == 0:\n",
    "                freqs_list_raw_s = GetLRTListByRow(lrtFile_for_s_0, 1)\n",
    "            else:\n",
    "                freqs_list_raw_s = GetLRTListFreq(lrtFile, s_ABC_round)\n",
    "            \n",
    "            LRT_table_s_het = []\n",
    "            LRT_table_s_common = []\n",
    "            LRT_table_s_bins = []\n",
    "\n",
    "            # Get summary statistics from allele frequencies\n",
    "            for freq_string in freqs_list_raw_s:\n",
    "\n",
    "                obs_het_s, obs_common_s, obs_bins_s = GetSummStats(freq_string, num_bins)\n",
    "                LRT_table_s_het.append(obs_het_s) \n",
    "                LRT_table_s_common.append(obs_common_s) \n",
    "                LRT_table_s_bins.append(obs_bins_s)\n",
    "\n",
    "            # Perform LRT\n",
    "            likelihood_0, likelihood_s_ABC, LR, LogLR, pval = LikelihoodRatioTest(LRT_table_0_het, \\\n",
    "                                    LRT_table_s_het, LRT_table_0_common, LRT_table_s_common, \\\n",
    "                                    LRT_table_0_bins, LRT_table_s_bins, LRT_num_sims, \\\n",
    "                                    obs_het, obs_common, obs_bins, constant_het, denom_het, \\\n",
    "                                    constant_common, denom_common, eps_bins, use_het, use_common, use_bins)\n",
    "\n",
    "            print(str(round(likelihood_0, 7)) + '\\t' + str(round(likelihood_s_ABC, 7)) + '\\t' + \\\n",
    "                              str(round(LR, 7)) + '\\t' + str(round(LogLR ,7)) + '\\t' + str(round(pval, 7)) + '\\n')\n",
    "\n",
    "    allele_freqs_file.close()\n",
    "            \n",
    "    \n",
    "    # Output statistics\n",
    "    print(\"Total lines: \" + str(total_lines) + \"\\n\")\n",
    "    percent_acc = total_lines_acc/total_lines\n",
    "    print(\"Total accepted lines: \" + str(total_lines_acc) + \" \" + str(percent_acc) + \"\\n\")\n",
    "    percent_0 = lower_0/total_lines\n",
    "    print(\"Lower bound for s is zero: \" + str(lower_0) + \" \" + str(percent_0) + \"\\n\")\n",
    "    \n",
    "    for i in range(0, len(s_acc)):\n",
    "        s_acc[i] = s_acc[i]/10000 * 100\n",
    "    \n",
    "    s_acc_med = np.median(s_acc)\n",
    "    s_acc_mean = np.mean(s_acc)\n",
    "\n",
    "    print(\"s_acc median: \" + str(s_acc_med) + \"\\n\")\n",
    "    print(\"s_acc mean: \" + str(s_acc_mean) + \"\\n\")\n",
    " \n",
    "    plt.figure(1)\n",
    "    buckets = list(np.arange(0, 101, 1)) \n",
    "    \n",
    "    plt.hist(s_acc, bins=buckets, weights=np.ones(len(s_acc)) / len(s_acc))\n",
    "\n",
    "    plt.title('Percent s accepted during ABC %s model \\n const_het %.5f, denom_het = %d, num_bins = %d \\n Summ stats used: het = %s common alleles = %s bins = %s'%(abc_model, constant_het, denom_het, num_bins, use_het, use_common, use_bins))\n",
    "    plt.xlabel(\"% s accepted\")\n",
    "    plt.ylabel(\"Frequency\")\n",
    "    #plt.savefig(figFile, bbox_inches='tight')\n",
    "    '''\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
