{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains code to plot heatmaps to explore feasible mutation rates consistent with observed allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "import os\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "PLOTDIR = 'pdfs-round1'\n",
    "PREFIX = \"Figure2-\"\n",
    "\n",
    "import glob\n",
    "import matplotlib.ticker as ticker\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Different colormap options: 'binary', 'hot_r', 'Blues', 'YlGnBu', 'Greens', 'viridis', 'cool'\n",
    "def main():\n",
    "    print('Running main')\n",
    "    combos = [(2, 'intergenic'),(3, 'intergenic'),(4, 'intergenic')] \n",
    "    superfolder = 'results_intergenic_top_x'\n",
    "    subfolder = '1kg_bin_agnostic_071921'\n",
    "    colormap = 'Blues'\n",
    "    folders_dict = {} # Key: Tuple (Period, whether intergenic/all loci), Value: List of folders corresponding to period\n",
    "    \n",
    "    folders_dict[(2, 'intergenic')] = ['t1','t2','t3','t4','t5','t6']\n",
    "    folders_dict[(3, 'intergenic')] = ['t7','t8','t9','t10','t11','t12','t13']\n",
    "    folders_dict[(4, 'intergenic')] =['t14','t15','t16','t17','t18','t19','t20']\n",
    "    \n",
    "    \n",
    "    num_files_dict = {} # Key: Period, Value: Number of files\n",
    "    num_files_dict[2] = 30\n",
    "    num_files_dict[3] = 72\n",
    "    num_files_dict[4] = 36\n",
    "    \n",
    "    opt_dic = {}\n",
    "    opt_dic[2] = ['11','12','13','14','15','16','17','18','19','20']\n",
    "    opt_dic[3] = ['05','06','07','08','09','10','11','12','13']\n",
    "    opt_dic[4] = ['07','08','09','10']\n",
    "    \n",
    "    motif_dic = {}\n",
    "    motif_dic[2] = ['AT','AG','AC']\n",
    "    motif_dic[3] = ['AAT','ATC','AAG','AGG','AGC','CCG','AAC','ACC'] \n",
    "    motif_dic[4] = ['AAAT','AATC','AATG','ATCC','AAAG','AAGG','AGAT','AAAC','ACAT']\n",
    "    \n",
    "    mut_setting_dic = {}\n",
    "    mut_setting_dic[2] = ['a','b','c','d','e','f']\n",
    "    mut_setting_dic[3] = ['a','b','c','d','e','f','g']\n",
    "    mut_setting_dic[4] = ['a','b','c','d','e','f','g']\n",
    "    \n",
    "    file_name_dic = {}\n",
    "    file_name_dic[2] = 'dinuc'\n",
    "    file_name_dic[3] = 'trinuc'\n",
    "    file_name_dic[4] = 'tetranuc'\n",
    "    \n",
    "    for combo in combos:\n",
    "        \n",
    "        per = combo[0]\n",
    "        list_of_percents = []\n",
    "        list_of_num_loci = []\n",
    "\n",
    "        # Plot heatmap for each motif in given period\n",
    "        for i in range(0,1):\n",
    "            for motif in motif_dic[per]:\n",
    "                for opt in opt_dic[per]:\n",
    "                    file_list = glob.glob('/gymreklab-tscc/bonnieh/joint_method/%s/%s/per_%d_*%s*opt_%s.txt'%(superfolder, subfolder, per, motif, opt))\n",
    "                    #file_list.sort() \n",
    "\n",
    "                    #print((motif,opt))\n",
    "                    '''\n",
    "                    if len(file_list) == 0:\n",
    "                        print('File truncated')\n",
    "                        print((folder, motif, opt))\n",
    "                        list_of_percents.append(0)\n",
    "                        if folder == folders_dict[combo][0]:\n",
    "                            list_of_num_loci.append(0)\n",
    "                    '''\n",
    "                    if 1>0:\n",
    "                        filename = file_list[0]\n",
    "                        results_file = open(filename, 'r')\n",
    "                        #lines = results_file.readlines()\n",
    "                        for k in range(0, 2):\n",
    "                            line= results_file.readline()\n",
    "                        \n",
    "                        info = line.strip().split(': ')\n",
    "                        num_loci = info[1]\n",
    "                        num_loci = num_loci.strip().split(' ')\n",
    "                        num_loci = int(num_loci[0])\n",
    "                        #print((motif,opt))\n",
    "                        #print(num_loci)\n",
    "                        list_of_num_loci.append(num_loci)\n",
    "                        #print((num_loci, motif, opt))\n",
    "                        if num_loci == 0:\n",
    "                            #print('0 loci')\n",
    "                            #print((folder, motif, opt))\n",
    "                            for elem in mut_setting_dic[per]:\n",
    "                                #list_of_percents.append(0) # np.log10(\n",
    "                                list_of_percents.append(float('-inf'))\n",
    "                        \n",
    "                        else:\n",
    "                            for k in range(0, 4):\n",
    "                                line= results_file.readline()\n",
    "                            info = line.strip().split(': ')\n",
    "                            list_settings = info[1]\n",
    "                            list_settings = list_settings.strip().split(', ')\n",
    "                            \n",
    "                            for elem in mut_setting_dic[per]:\n",
    "                                num_acc = list_settings.count(elem)\n",
    "                                \n",
    "                                if num_acc == 0:\n",
    "                                    list_of_percents.append(float('-inf'))\n",
    "                                else:\n",
    "                                    list_of_percents.append(num_acc/500)\n",
    "                            \n",
    "                        results_file.close()\n",
    "        matrix = np.zeros((len(folders_dict[combo]), num_files_dict[per]))\n",
    "        #print(matrix)\n",
    "        counter = 0\n",
    "\n",
    "        for i in range(num_files_dict[per]): \n",
    "            for j in range(len(mut_setting_dic[per])):\n",
    "                #print(i,j)\n",
    "                matrix[j,i] = list_of_percents[counter]\n",
    "                counter = counter + 1\n",
    "\n",
    "        # Visualize heatmap\n",
    "\n",
    "        #print(matrix)\n",
    "        if per == 2:\n",
    "            settings = [\"a\", \"b\", \"c\", \"d\",\"e\",\"f\"] # Mutation rate settings for period 2, used to start with z and end with e\n",
    "            \n",
    "        else:\n",
    "            settings = ['a','b','c','d','e','f','g'] #['a','b','bc','c','cd','d','de','e','f','g'] # Orig up to d, Mutation rate settings for periods 3 and 4\n",
    "\n",
    "\n",
    "        opt = []\n",
    "        for mot in motif_dic[per]:\n",
    "            for opt_num in opt_dic[per]:\n",
    "                if opt_num[0] == '0':\n",
    "                    opt.append(opt_num[1]) #mot + ' ' \n",
    "                else:\n",
    "                    opt.append(opt_num)\n",
    "\n",
    "        # Plot\n",
    "        \n",
    "        if combo[0] == 2:\n",
    "            fig, axs = plt.subplots(2, 1, gridspec_kw={'height_ratios': [1, 3]},sharex=True,figsize=(15,7))\n",
    "        if combo[0] == 3:\n",
    "            fig, axs = plt.subplots(2, 1, gridspec_kw={'height_ratios': [1, 3]},sharex=True,figsize=(15,7)) # 3\n",
    "        if combo[0] == 4:\n",
    "            fig, axs = plt.subplots(2, 1, gridspec_kw={'height_ratios': [1, 3]},sharex=True,figsize=(15,7)) # 5\n",
    "        \n",
    "        ax1 = axs[0]\n",
    "        ax2 = axs[1]\n",
    "        \n",
    "        y_pos = list(range(0,len(list_of_num_loci)))\n",
    "        ax1.bar(y_pos, list_of_num_loci)  \n",
    "           \n",
    "        ax1.set_ylabel(\"Number of loci\",size=15)\n",
    "        ax1.tick_params(axis=\"y\", labelsize=15)\n",
    "       \n",
    "        #print(matrix)\n",
    "        im = ax2.imshow(matrix,cmap=colormap)#np.log10(\n",
    "        \n",
    "        #cbar = fig.colorbar(im)\n",
    "        #cbar.ax.set_ylabel('Log 10 Number of ABC Acceptances')\n",
    "        #cbar.ax.set_ylabel('Proportion of ABC Acceptances')\n",
    "\n",
    "        # Show all ticks\n",
    "        ax2.set_xticks(np.arange(len(opt)))\n",
    "        ax2.set_yticks(np.arange(len(settings)))\n",
    "\n",
    "        # Label ticks\n",
    "        ax2.set_xticklabels(opt)\n",
    "        ax2.set_yticklabels(settings)\n",
    "\n",
    "        # Rotate tick labels \n",
    "        plt.setp(ax2.get_xticklabels(), rotation=45, ha=\"right\", rotation_mode=\"anchor\")\n",
    "        plt.setp(ax2.get_yticklabels(), ha=\"right\", rotation_mode=\"anchor\")\n",
    "        if per == 2:\n",
    "            ax2.set_ylim(5.5,-0.5) # x=5.5 for period 2, x=3.5 for periods 3 and 4\n",
    "        else:\n",
    "            ax2.set_ylim(6.5,-0.5) # 3.5\n",
    "        '''\n",
    "        # Loop over data and create text annotations\n",
    "        for i in range(len(settings)):\n",
    "            for j in range(len(opt)):\n",
    "                text = ax.text(j, i, transition_matrix[i, j],\n",
    "                               ha=\"center\", va=\"center\", color=\"w\")\n",
    "        '''\n",
    "        #ax.set_title(\"Dependence of ABC acceptance rates \\n on mutation rates\")\n",
    "        if combo[0] == 2:\n",
    "            ax1.set_title('Dinucleotides',fontsize=20)\n",
    "        if combo[0] == 3:\n",
    "            ax1.set_title('Trinucleotides',fontsize=20)\n",
    "        if combo[0] == 4:\n",
    "            ax1.set_title('Tetranucleotides',fontsize=20)\n",
    "        \n",
    "        # Position x axis label\n",
    "        if combo[0] == 3:\n",
    "            plt.text(len(motif_dic[per])*len(opt_dic[per])/2-5,len(settings)+4, 'Motif and Optimal allele',size=15)\n",
    "        else:\n",
    "            plt.text(len(motif_dic[per])*len(opt_dic[per])/2-3,len(settings)+2, 'Motif and Optimal allele',size=15)\n",
    "        ax2.set_ylabel('Mutation rate\\nsetting',size=15)\n",
    "        \n",
    "        count = 0\n",
    "        for motif in motif_dic[per]:\n",
    "            count = count + 1\n",
    "            num_opt_alleles = len(opt_dic[per])\n",
    "            plt.axvline(x=count*num_opt_alleles-0.5, linestyle = '-',c='black')\n",
    "            if combo[0] == 3:\n",
    "                plt.text(count*num_opt_alleles-1-(num_opt_alleles/2),len(settings)+2.5,motif,size=15)\n",
    "            else:\n",
    "                plt.text(count*num_opt_alleles-1-(num_opt_alleles/2),len(settings)+1,motif,size=15)\n",
    "        plt.savefig(os.path.join(PLOTDIR + PREFIX+'heatmap_%d_%s_top_x.pdf'%(per, combo[1])),bbox_inches='tight')\n",
    "        plt.yticks(size=15)\n",
    "        plt.xticks(size=15)\n",
    "        fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running main\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
