{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains code to measure the burden of deleterious variation from STRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/site-packages/IPython/core/magics/pylab.py:160: UserWarning: pylab import has clobbered these variables: ['sample', 'info', 'copy']\n",
      "`%matplotlib` prevents importing * from pylab and numpy\n",
      "  \"\\n`%matplotlib` prevents importing * from pylab and numpy\"\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "from matplotlib import pyplot as plt\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "sys.path.append(\"/storage/BonnieH/selection_project/helper_functions\")\n",
    "from Simulation_functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# New mutations - based on models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each class of STRs, we want:\n",
    "# - motif/opt allele\n",
    "# - number of STRs\n",
    "# - mutation rate\n",
    "# - mean s per mutation\n",
    "# Then will sum across all categories\n",
    "\n",
    "# Read in joint results\n",
    "joint_results = {}\n",
    "sistr2_scores_file = open('/storage/BonnieH/selection_project/analysis/SISTR2_scores_by_motif_corrected_num_loci.txt', 'r')\n",
    "sistr2_scores_file.readline()\n",
    "for line in sistr2_scores_file:\n",
    "    info = line.strip().split('\\t')\n",
    "    period = int(info[0])\n",
    "    optimal_ru = int(info[1])\n",
    "    motif = info[2]\n",
    "    num_loci = int(info[3])\n",
    "    est_a = float(info[4])\n",
    "    est_b = float(info[5])\n",
    "    joint_results[(period,optimal_ru,motif,num_loci)] = (est_a,est_b)\n",
    "sistr2_scores_file.close()\n",
    "\n",
    "# List contents: mu, beta, p, l, optimal ru for the mu value\n",
    "# Mutation model parameters for different motifs\n",
    "motif_info = {}\n",
    "motif_info['eurodem_prior2_dinuc_e_1kg_euro'] = [10**-5, 0.3, 0.6, 0.15, 6]\n",
    "motif_info['eurodem_prior2_dinuc_d_1kg_euro'] = [10**-4.6, 0.3, 0.6, 0.135, 6]\n",
    "motif_info['eurodem_prior2_trinuc_e_1kg_euro'] = [10**-6, 0.3, 0.9, 0.3, 5] \n",
    "motif_info['eurodem_prior2_tetranuc_b_1kg_euro'] = [10**-5, 0.3, 0.9, 0.25, 3]\n",
    "motif_info['eurodem_prior2_tetranuc_c_1kg_euro'] = [10**-5.5, 0.3, 0.9, 0.42, 3]\n",
    "motif_info['eurodem_prior2_tetranuc_d_1kg_euro'] = [10**-6, 0.3, 0.9, 0.45, 3]\n",
    "\n",
    "# Mutation model settings for different motifs (inferred by SISTR2)\n",
    "trinuc_motifs = ['ACC','AGG','AGC','CCG','AAG','AAC','ATC','AAT']\n",
    "dinuc_motifs  = ['AC','AG','AT'] \n",
    "tetranuc_motifs = ['AAGG','AAAT','ATCC','AATC','AAAC','AATG','ACAT','AGAT','AAAG'] \n",
    "mut_setting = {}\n",
    "mut_setting['AC'] = 'eurodem_prior2_dinuc_e_1kg_euro'\n",
    "mut_setting['AG'] = 'eurodem_prior2_dinuc_e_1kg_euro'\n",
    "mut_setting['AT'] = 'eurodem_prior2_dinuc_d_1kg_euro'\n",
    "for mot in trinuc_motifs:\n",
    "    mut_setting[mot] = 'eurodem_prior2_trinuc_e_1kg_euro'\n",
    "for mot in ['AAAG', 'AAGG', 'AGAT']:\n",
    "    mut_setting[mot] = 'eurodem_prior2_tetranuc_b_1kg_euro'\n",
    "for mot in ['ACAT', 'AAAT']:\n",
    "    mut_setting[mot] = 'eurodem_prior2_tetranuc_c_1kg_euro'\n",
    "for mot in ['AATC', 'AATG', 'ATCC', 'AAAC']:\n",
    "    mut_setting[mot] = 'eurodem_prior2_tetranuc_d_1kg_euro'\n",
    "\n",
    "    \n",
    "motifs_ = []\n",
    "opts_ = []\n",
    "numloci_ = []\n",
    "means_ = []\n",
    "muts_ = []\n",
    "# Get burden for each category\n",
    "for elem in joint_results:\n",
    "    # Get params for this class\n",
    "    setting_info = mut_setting[elem[2]]\n",
    "    mut_info = motif_info[setting_info]\n",
    "    num_loci = elem[3] # target size\n",
    "    \n",
    "    # Get mutation rate   \n",
    "    log_mu_prime = np.log10(mut_info[0])+mut_info[3]*(elem[1] - mut_info[4])\n",
    "    mu_prime = 10**log_mu_prime\n",
    "    \n",
    "    if mu_prime < 10**-8: mu_prime = 10**-8 \n",
    "    if mu_prime > 10**-3: mu_prime = 10**-3\n",
    "\n",
    "    # Get mean s value of a new mutation\n",
    "    # Sum over all possible step sizes\n",
    "    mean_s = 0\n",
    "    for i in range(-20, 20):\n",
    "        if elem[1] + i <= 0: continue # can't have negative repeat copy nmber\n",
    "        prob_step = GetStepSizeProb(0, i, mut_info[1], mut_info[2])\n",
    "        sval = joint_results[elem][0]*joint_results[elem][1]*abs(i) # a*b*step\n",
    "        mean_s += prob_step*sval\n",
    "    \n",
    "    motifs_.append(elem[2])\n",
    "    opts_.append(elem[1])\n",
    "    numloci_.append(num_loci)\n",
    "    means_.append(mean_s)\n",
    "    muts_.append(mu_prime)\n",
    "\n",
    "data = pd.DataFrame({\"motif\": motifs_,\n",
    "             \"opt\": opts_,\n",
    "             \"numloci\": numloci_,\n",
    "             \"mean.s\": means_,\n",
    "             \"mu\": muts_})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.051146734481617315\n",
      "Num STRs: 74119\n",
      "Mean s/mutation: 0.0006540766876531336\n",
      "Mean mu: 0.00019606154782613472\n",
      "Overall burden: 0.01900993554167439\n"
     ]
    }
   ],
   "source": [
    "# Get burden\n",
    "\n",
    "data[\"burden\"] = data.apply(lambda x: 2*x[\"mu\"]*x[\"mean.s\"]*x[\"numloci\"], 1)\n",
    "data.to_csv(\"figure4-burden-numbers.csv\")\n",
    "print(np.sum(data[\"burden\"]))\n",
    "\n",
    "# Summarize overall\n",
    "num_strs = np.sum(data[\"numloci\"])\n",
    "mean_s = np.sum(data.apply(lambda x: x[\"numloci\"]*x[\"mean.s\"], 1))/np.sum(data[\"numloci\"])\n",
    "mean_mu = np.sum(data.apply(lambda x: x[\"numloci\"]*x[\"mu\"], 1))/np.sum(data[\"numloci\"])\n",
    "burden = 2*mean_mu*num_strs*mean_s\n",
    "print(\"Num STRs: %s\"%(num_strs))\n",
    "print(\"Mean s/mutation: %s\"%(mean_s))\n",
    "print(\"Mean mu: %s\"%(mean_mu))\n",
    "print(\"Overall burden: %s\"%burden)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# New mutations - based on observed de novos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "joint_results2 = {} # motif, opt -> s\n",
    "for elem in joint_results:\n",
    "    joint_results2[(elem[2], elem[1])] = joint_results[elem][0]*joint_results[elem][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### NA12864 #####\n",
      "Num mutations: 71\n",
      "Num mutations neg - perlocus: 28\n",
      "Num mutations neg - sistr2: 44\n",
      "Total burden (perlocus): 0.10155 (0.011636000000000002-0.2417424999999997)\n",
      "Total burden (sistr2): 0.04259394431339999 (0.0008853290558367549-0.07949122466458347)\n",
      "Total burden only neg (perlocus): 0.14090000000000005 (0.05164875000000001-0.10839943973711944)\n",
      "Total burden only neg (sistr2): 0.07377828493375 (0.044411613476670504-0.10839943973711944)\n",
      "##### NA10865 #####\n",
      "Num mutations: 84\n",
      "Num mutations neg - perlocus: 34\n",
      "Num mutations neg - sistr2: 55\n",
      "Total burden (perlocus): -0.004800000000000003 (-0.05077700000000002-0.031197999999999938)\n",
      "Total burden (sistr2): 0.01597992023192 (-0.012781489923731248-0.04794682284641523)\n",
      "Total burden only neg (perlocus): 0.035710000000000006 (0.023088499999999998-0.06398801371731348)\n",
      "Total burden only neg (sistr2): 0.042243841245500006 (0.026944061604484246-0.06398801371731348)\n",
      "##### NA10845 #####\n",
      "Num mutations: 22\n",
      "Num mutations neg - perlocus: 9\n",
      "Num mutations neg - sistr2: 15\n",
      "Total burden (perlocus): 0.016470000000000002 (-0.00643025-0.042253)\n",
      "Total burden (sistr2): 0.00880774667467 (-0.004309949200254749-0.020150850677598243)\n",
      "Total burden only neg (perlocus): 0.02357 (0.00370925-0.02722713651266849)\n",
      "Total burden only neg (sistr2): 0.01732052853815 (0.009176821102907252-0.02722713651266849)\n"
     ]
    }
   ],
   "source": [
    "NUMBS = 100\n",
    "\n",
    "# Load SISTR scores\n",
    "def GetSEst(x):\n",
    "    if x[\"LRT_p_value\"]>1:\n",
    "        return 0\n",
    "    else: return x[\"ABC_s_median\"]\n",
    "    \n",
    "sistr_scores_file = \"/storage/BonnieH/selection_project/1000genomes/euro_allele_freqs/SISTR_per_locus_scores.txt\"\n",
    "sistr = pd.read_csv(sistr_scores_file, sep=\"\\t\")\n",
    "sistr[\"chrom\"] = sistr[\"chrom\"].apply(lambda x: \"chr\"+str(x))\n",
    "sistr[\"s.est\"] = sistr.apply(GetSEst,1)\n",
    "sistr[\"sistr2\"] = sistr.apply(lambda x: joint_results2.get((x[\"motif\"], x[\"optimal_ru\"]), None), 1)\n",
    "\n",
    "# Load all de novos + filter\n",
    "denovos_file = \"/storage/hziaeija/variant_trio/trio_variants/monstr_all_filters/CEU_all_filters.csv\"\n",
    "denovo = pd.read_csv(denovos_file)\n",
    "denovo = denovo[(denovo[\"encl_mother\"]==0) & (denovo[\"encl_father\"]==0) & (denovo[\"encl_child\"]>=5)]\n",
    "denovo[\"start\"] = denovo[\"pos\"]\n",
    "\n",
    "def BootstrapSVals(all_svals, all_svals_2):\n",
    "    bs_svals = []\n",
    "    bs_svals_2 = []\n",
    "    for i in range(len(all_svals)):\n",
    "        ind = random.randint(0, len(all_svals))\n",
    "        bs_svals.append(all_svals[ind])\n",
    "        bs_svals_2.append(all_svals_2[ind])\n",
    "    return bs_svals, bs_svals_2\n",
    "\n",
    "def GetLowerBS(vals):\n",
    "    return np.percentile(vals, [2.5])[0] #np.min(vals)\n",
    "\n",
    "def GetUpperBS(vals):\n",
    "    return np.percentile(vals, [97.5])[0] #np.max(vals)\n",
    "\n",
    "for sample in [\"NA12864\", \"NA10865\",\"NA10845\" ]:\n",
    "    denovo_ = denovo[denovo[\"child\"]==sample].copy()\n",
    "    ddata = pd.merge(denovo_[[\"chrom\",\"start\",\"newallele\",\"mutsize\"]], sistr, on=[\"chrom\",\"start\"])\n",
    "    ddata = ddata[~np.isnan(ddata[\"ABC_s_median\"]) & ~np.isnan(ddata[\"sistr2\"])]\n",
    "\n",
    "    ddata[\"s.new\"] = ddata.apply(lambda x: (x[\"newallele\"]-x[\"optimal_ru\"])*x[\"s.est\"], 1)\n",
    "    ddata[\"s.old\"] = ddata.apply(lambda x: (x[\"newallele\"]-x[\"mutsize\"]-x[\"optimal_ru\"])*x[\"s.est\"], 1)\n",
    "    ddata[\"s.delta\"] = ddata[\"s.new\"]-ddata[\"s.old\"]\n",
    "\n",
    "    ddata[\"s2.new\"] = ddata.apply(lambda x: (x[\"newallele\"]-x[\"optimal_ru\"])*x[\"sistr2\"], 1)\n",
    "    ddata[\"s2.old\"] = ddata.apply(lambda x: (x[\"newallele\"]-x[\"mutsize\"]-x[\"optimal_ru\"])*x[\"sistr2\"], 1)\n",
    "    ddata[\"s2.delta\"] = ddata[\"s2.new\"]-ddata[\"s2.old\"]\n",
    "\n",
    "    print (\"##### %s #####\"%sample)\n",
    "    print(\"Num mutations: %s\"%ddata.shape[0])\n",
    "    print(\"Num mutations neg - perlocus: %s\"%ddata[ddata[\"s.delta\"]>0].shape[0])\n",
    "    print(\"Num mutations neg - sistr2: %s\"%ddata[ddata[\"s2.delta\"]>0].shape[0])\n",
    "\n",
    "    # Get bootstrap CIs for all mutations, only negative - TODO\n",
    "    all_svals = list(ddata[\"s.delta\"])\n",
    "    all_svals_2 = list(ddata[\"s2.delta\"])\n",
    "    \n",
    "    burden_bs = []\n",
    "    burden_bs_2 = []\n",
    "    burden_neg_bs = []\n",
    "    burden_neg_bs_2 = []\n",
    "    \n",
    "    # Bootstrap sample mutations\n",
    "    for i in range(NUMBS):\n",
    "        bs_svals, bs_svals_2 = BootstrapSVals(all_svals, all_svals_2)\n",
    "        burden_bs.append(np.sum(bs_svals))\n",
    "        burden_bs_2.append(np.sum(bs_svals_2))\n",
    "        burden_neg_bs.append(np.sum([item for item in bs_svals if item > 0]))\n",
    "        burden_neg_bs_2.append(np.sum([item for item in bs_svals_2 if item > 0]))\n",
    "    \n",
    "    print(\"Total burden (perlocus): %s (%s-%s)\"%(np.sum(ddata[\"s.delta\"]), GetLowerBS(burden_bs), GetUpperBS(burden_bs)))\n",
    "    print(\"Total burden (sistr2): %s (%s-%s)\"%(np.sum(ddata[\"s2.delta\"]), GetLowerBS(burden_bs_2), GetUpperBS(burden_bs_2)))\n",
    "    print(\"Total burden only neg (perlocus): %s (%s-%s)\"%(np.sum(ddata[ddata[\"s.delta\"]>0][\"s.delta\"]), GetLowerBS(burden_neg_bs), GetUpperBS(burden_neg_bs_2)))\n",
    "    print(\"Total burden only neg (sistr2): %s (%s-%s)\"%(np.sum(ddata[ddata[\"s2.delta\"]>0][\"s2.delta\"]), GetLowerBS(burden_neg_bs_2), GetUpperBS(burden_neg_bs_2)))\n",
    "\n",
    "    #ddata[[\"chrom\",\"start\",\"period\",\"motif\",\"newallele\",\"mutsize\",\"ABC_s_median\",\"s.est\",\"optimal_ru\",\"s.old\",\"s.new\",\"s.delta\",\"LRT_p_value\"]].sort_values(\"s.delta\", ascending=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
