{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook coontains code to assess goodness of fit of the SISTR2 motif analysis (a,b) scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "import sys\n",
    "sys.path.append(\"/gymreklab-tscc/bonnieh/helper_functions\")\n",
    "from Joint_method_functions import *\n",
    "import glob\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "import matplotlib.ticker as ticker\n",
    "import seaborn as sns\n",
    "from statistics import mode\n",
    "from scipy import stats\n",
    "\n",
    "PLOTDIR = '/storage/BonnieH/selection_project/figures/supp_figures/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given file, get ABC lookup table\n",
    "def getABCTable(file):\n",
    "    num_bins = 0\n",
    "    \n",
    "    # Return list of lists that contain [s, het, common]\n",
    "    table = GetABCList(file, num_bins)\n",
    "    dic_summ_stats = {}\n",
    "        \n",
    "    # Fill in dic_summ_stats: Key is s, value is list of het, number of common alleles pairs for given s value\n",
    "    for combo in table:\n",
    "        s_round = get_LRT_bin(combo[0])\n",
    "        if s_round not in dic_summ_stats:\n",
    "            dic_summ_stats[s_round] = []\n",
    "        dic_summ_stats[s_round].append([combo[1], combo[2]]) # Append het, number of common alleles for s value\n",
    "    \n",
    "    return dic_summ_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Different colormap options: 'binary', 'hot_r', 'Blues', 'YlGnBu', 'Greens', 'viridis', 'cool'\n",
    "def main():\n",
    "    \n",
    "    pval_threshold = 0.05\n",
    "    goodness_of_fit_file = open('goodness_of_fit_motif_analysis.txt','w')\n",
    "    \n",
    "    #print('Running main')\n",
    "    combos = [(2, 'intergenic'),(3, 'intergenic'),(4, 'intergenic')]\n",
    "    superfolder = 'results_top_x'\n",
    "    subfolder = 'motif_analysis_diff_mu_071921'\n",
    "    colormap = 'Greens'\n",
    "    \n",
    "    num_files_dict = {} # Key: Period, Value: Number of files\n",
    "    num_files_dict[2] = 30\n",
    "    num_files_dict[3] = 72\n",
    "    num_files_dict[4] = 36\n",
    "    \n",
    "    opt_dic = {}\n",
    "    opt_dic[2] = ['11','12','13','14','15','16','17','18','19','20']\n",
    "    opt_dic[3] = ['05','06','07','08','09','10','11','12','13']\n",
    "    opt_dic[4] = ['07','08','09','10']\n",
    "    \n",
    "    motif_dic = {}\n",
    "    motif_dic[2] = ['AT','AG','AC']\n",
    "    motif_dic[3] = ['AAT','ATC','AAG','AGG','AGC','CCG','AAC','ACC'] \n",
    "    motif_dic[4] = ['AAAT','AATC','AATG','ATCC','AAAG','AAGG','AGAT','AAAC','ACAT']\n",
    "    \n",
    "    file_name_dic = {}\n",
    "    file_name_dic[2] = 'dinuc'\n",
    "    file_name_dic[3] = 'trinuc'\n",
    "    file_name_dic[4] = 'tetranuc'\n",
    "    \n",
    "    for combo in combos:\n",
    "        \n",
    "        per = combo[0]\n",
    "        list_of_percents = []\n",
    "        list_of_num_loci = []\n",
    "        list_of_pvalue= []\n",
    "\n",
    "        # Plot heatmap for each motif in given period\n",
    "        for i in range(0,1):\n",
    "            for motif in motif_dic[per]:\n",
    "                for opt in opt_dic[per]:\n",
    "                    file_list = glob.glob('/gymreklab-tscc/bonnieh/joint_method/%s/%s/per_%d_*%s*opt_%s.txt'%(superfolder, subfolder, per, motif, opt))\n",
    "                    \n",
    "                    if 1>0:\n",
    "                        filename = file_list[0]\n",
    "                        results_file = open(filename, 'r')\n",
    "                        \n",
    "                        for k in range(0, 2):\n",
    "                            line= results_file.readline()\n",
    "                        \n",
    "                        info = line.strip().split(': ')\n",
    "                        num_loci = info[1]\n",
    "                        num_loci = num_loci.strip().split(' ')\n",
    "                        num_loci = int(num_loci[0])\n",
    "                        \n",
    "                        list_of_num_loci.append(num_loci)\n",
    "                        \n",
    "                        if num_loci == 0:\n",
    "                            \n",
    "                            for elem in mut_setting_dic[per]:\n",
    "                                list_of_percents.append(0)\n",
    "                                list_of_pvalue.append(0)\n",
    "                        \n",
    "                        else:\n",
    "                            \n",
    "                            if opt[0] == '0':\n",
    "                                opt_num=int(opt[1]) \n",
    "                            else:\n",
    "                                opt_num=int(opt)\n",
    "\n",
    "                            header = results_file.readline().strip().split('\\t')\n",
    "                            info = results_file.readline().strip().split(': ')\n",
    "\n",
    "                            list_het = info[1]\n",
    "                            list_het = list_het[1:]\n",
    "                            list_het = list_het[:-1]\n",
    "                            het_list = list_het.strip().split(', ')\n",
    "\n",
    "                            obs_het_distr = []\n",
    "                            opt_allele_sub_list = []\n",
    "                            for elem in het_list:\n",
    "                                obs_het_distr.append(float(elem))\n",
    "                                opt_allele_sub_list.append((per, opt_num))\n",
    "\n",
    "                            for k in range(0, 3):\n",
    "                                line= results_file.readline()\n",
    "                                \n",
    "                            info = line.strip().split(',')\n",
    "                            a_val = float(info[0])\n",
    "                            b_val = float(info[1])\n",
    "                            #print('a val')\n",
    "                            #print(a_val)\n",
    "                            #print('b val')\n",
    "                            #print(b_val)\n",
    "                            \n",
    "                            #print((motif,opt))\n",
    "                            mode_setting = filename.split('_')\n",
    "                            #print(mode_setting)\n",
    "                            mode_setting = mode_setting[13]\n",
    "                            #print('mode setting')\n",
    "                            #print(mode_setting)\n",
    "                                \n",
    "                            goodness_of_fit_file.write(motif + '\\t' + str(opt_num) + '\\t' + mode_setting + '\\n')\n",
    "                            #obs_het_distr, opt_allele_sub_list = GetObsHetDistrFn(motif, opt_num)\n",
    "                            \n",
    "                            # Get simulated heterozygosity\n",
    "                            \n",
    "                            mut_setting_folder_name = {}\n",
    "                            mut_setting_folder_name[2] = 'eurodem_prior2_dinuc_'\n",
    "                            mut_setting_folder_name[3] = 'eurodem_prior2_trinuc_'\n",
    "                            mut_setting_folder_name[4] = 'eurodem_prior2_tetranuc_'\n",
    "\n",
    "                            folder_prefix = mut_setting_folder_name[len(motif)]\n",
    "                            \n",
    "                            file = '/gymreklab-tscc/bonnieh/abc/results/' + folder_prefix + mode_setting + '_1kg_euro/' + str(len(motif)) + '_' + str(opt_num) + '.txt'\n",
    "\n",
    "                            ABC_dic = {}\n",
    "                            ABC_dic[opt_num] = getABCTable(file)\n",
    "                            \n",
    "                            ABC_tables_mut = ABC_dic\n",
    "                            count = 0\n",
    "                            for zz in range(0, 100):\n",
    "                                sim_het_list, sim_common_list = GetLists(ABC_tables_mut, opt_allele_sub_list, a_val, b_val)\n",
    "                                if len(obs_het_distr) != len(sim_het_list):\n",
    "                                    print('Observed and simulated het different lengths')\n",
    "                                ks_stat, pval = stats.ks_2samp(obs_het_distr, sim_het_list)\n",
    "                                if pval > pval_threshold:\n",
    "                                    count = count + 1\n",
    "                                #if pval < 0.05:\n",
    "                                    #list_of_pvalue.append(0)\n",
    "                                #else:\n",
    "                                    #list_of_pvalue.append(1)\n",
    "                            list_of_pvalue.append(count)\n",
    "                            print(\"%s\\t%s\\t%s\"%(motif, opt, count))\n",
    "                            #print(count)\n",
    "                            \n",
    "                        results_file.close()\n",
    "        matrix = np.zeros((len(motif_dic[per]), len(opt_dic[per])))\n",
    "        #print(matrix)\n",
    "        counter = 0\n",
    "\n",
    "        for i in range(len(motif_dic[per])): \n",
    "            for j in range(len(opt_dic[per])):\n",
    "                #print(i,j)\n",
    "                matrix[i,j] = list_of_pvalue[counter]\n",
    "                counter = counter + 1\n",
    "\n",
    "        # Visualize heatmap\n",
    "\n",
    "        #print(matrix)\n",
    "      \n",
    "        opt = []\n",
    "        \n",
    "        for opt_num in opt_dic[per]:\n",
    "            if opt_num[0] == '0':\n",
    "                opt.append(opt_num[1]) \n",
    "            else:\n",
    "                opt.append(opt_num)\n",
    "\n",
    "        print(matrix)\n",
    "        \"\"\"\n",
    "        fig, ax = plt.subplots()\n",
    "        \n",
    "        im = ax.imshow(matrix,cmap=colormap) \n",
    "        cbar = fig.colorbar(im)\n",
    "        cbar.ax.set_ylabel('Percent simulations with good fit')\n",
    "\n",
    "        # Show all ticks\n",
    "        ax.set_xticks(np.arange(len(opt)))\n",
    "        ax.set_yticks(np.arange(len(motif_dic[per])))\n",
    "\n",
    "        # Label ticks\n",
    "        ax.set_xticklabels(opt)\n",
    "        ax.set_yticklabels(motif_dic[per])\n",
    "\n",
    "        if per == 2:\n",
    "            ax.set_title('Dinucleotides',fontsize=20)\n",
    "        if per == 3:\n",
    "            ax.set_title('Trinucleotides',fontsize=20)\n",
    "        if per == 4:\n",
    "            ax.set_title('Tetranucleotides',fontsize=20)\n",
    "        \n",
    "        plt.savefig(PLOTDIR + 'heatmap_%d_w_cbar.pdf'%(per),bbox_inches='tight')\n",
    "        plt.yticks(size=15)\n",
    "        plt.xticks(size=15)\n",
    "        fig.tight_layout()\n",
    "        \"\"\"\n",
    "    goodness_of_fit_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT\t11\t95\n",
      "AT\t12\t19\n",
      "AT\t13\t1\n",
      "AT\t14\t0\n",
      "AT\t15\t2\n",
      "AT\t16\t31\n",
      "AT\t17\t56\n",
      "AT\t18\t93\n",
      "AT\t19\t77\n",
      "AT\t20\t100\n",
      "AG\t11\t0\n",
      "AG\t12\t96\n",
      "AG\t13\t58\n",
      "AG\t14\t100\n",
      "AG\t15\t62\n",
      "AG\t16\t100\n",
      "AG\t17\t89\n",
      "AG\t18\t95\n",
      "AG\t19\t95\n",
      "AG\t20\t100\n",
      "AC\t11\t76\n",
      "AC\t12\t95\n",
      "AC\t13\t96\n",
      "AC\t14\t65\n",
      "AC\t15\t85\n",
      "AC\t16\t94\n",
      "AC\t17\t100\n",
      "AC\t18\t79\n",
      "AC\t19\t2\n",
      "AC\t20\t4\n",
      "[[ 95.  19.   1.   0.   2.  31.  56.  93.  77. 100.]\n",
      " [  0.  96.  58. 100.  62. 100.  89.  95.  95. 100.]\n",
      " [ 76.  95.  96.  65.  85.  94. 100.  79.   2.   4.]]\n",
      "AAT\t05\t0\n",
      "AAT\t06\t0\n",
      "AAT\t07\t0\n",
      "AAT\t08\t0\n",
      "AAT\t09\t0\n",
      "AAT\t10\t0\n",
      "AAT\t11\t0\n",
      "AAT\t12\t0\n",
      "AAT\t13\t0\n",
      "ATC\t05\t0\n",
      "ATC\t06\t0\n",
      "ATC\t07\t15\n",
      "ATC\t08\t99\n",
      "ATC\t09\t98\n",
      "ATC\t10\t100\n",
      "ATC\t11\t100\n",
      "ATC\t12\t53\n",
      "ATC\t13\t99\n",
      "AAG\t05\t50\n",
      "AAG\t06\t26\n",
      "AAG\t07\t0\n",
      "AAG\t08\t35\n",
      "AAG\t09\t78\n",
      "AAG\t10\t77\n",
      "AAG\t11\t100\n",
      "AAG\t12\t74\n",
      "AAG\t13\t100\n",
      "AGG\t05\t0\n",
      "AGG\t06\t0\n",
      "AGG\t07\t80\n",
      "AGG\t08\t100\n",
      "AGG\t09\t100\n",
      "AGG\t10\t99\n",
      "AGG\t11\t94\n",
      "AGG\t12\t97\n",
      "AGG\t13\t82\n",
      "AGC\t05\t0\n",
      "AGC\t06\t0\n",
      "AGC\t07\t22\n",
      "AGC\t08\t47\n",
      "AGC\t09\t98\n",
      "AGC\t10\t100\n",
      "AGC\t11\t99\n",
      "AGC\t12\t100\n",
      "AGC\t13\t100\n",
      "CCG\t05\t69\n",
      "CCG\t06\t0\n",
      "CCG\t07\t4\n",
      "CCG\t08\t20\n",
      "CCG\t09\t99\n",
      "CCG\t10\t97\n",
      "CCG\t11\t99\n",
      "CCG\t12\t96\n",
      "CCG\t13\t100\n",
      "AAC\t05\t35\n",
      "AAC\t06\t0\n",
      "AAC\t07\t0\n",
      "AAC\t08\t1\n",
      "AAC\t09\t12\n",
      "AAC\t10\t5\n",
      "AAC\t11\t25\n",
      "AAC\t12\t4\n",
      "AAC\t13\t99\n",
      "ACC\t05\t0\n",
      "ACC\t06\t0\n",
      "ACC\t07\t2\n",
      "ACC\t08\t36\n",
      "ACC\t09\t96\n",
      "ACC\t10\t100\n",
      "ACC\t11\t100\n",
      "ACC\t12\t100\n",
      "ACC\t13\t100\n",
      "[[  0.   0.   0.   0.   0.   0.   0.   0.   0.]\n",
      " [  0.   0.  15.  99.  98. 100. 100.  53.  99.]\n",
      " [ 50.  26.   0.  35.  78.  77. 100.  74. 100.]\n",
      " [  0.   0.  80. 100. 100.  99.  94.  97.  82.]\n",
      " [  0.   0.  22.  47.  98. 100.  99. 100. 100.]\n",
      " [ 69.   0.   4.  20.  99.  97.  99.  96. 100.]\n",
      " [ 35.   0.   0.   1.  12.   5.  25.   4.  99.]\n",
      " [  0.   0.   2.  36.  96. 100. 100. 100. 100.]]\n",
      "AAAT\t07\t83\n",
      "AAAT\t08\t100\n",
      "AAAT\t09\t37\n",
      "AAAT\t10\t22\n",
      "AATC\t07\t26\n",
      "AATC\t08\t96\n",
      "AATC\t09\t99\n",
      "AATC\t10\t100\n",
      "AATG\t07\t68\n",
      "AATG\t08\t80\n",
      "AATG\t09\t100\n",
      "AATG\t10\t99\n",
      "ATCC\t07\t99\n",
      "ATCC\t08\t98\n",
      "ATCC\t09\t100\n",
      "ATCC\t10\t79\n",
      "AAAG\t07\t98\n",
      "AAAG\t08\t25\n",
      "AAAG\t09\t75\n",
      "AAAG\t10\t60\n",
      "AAGG\t07\t100\n",
      "AAGG\t08\t19\n",
      "AAGG\t09\t94\n",
      "AAGG\t10\t37\n",
      "AGAT\t07\t2\n",
      "AGAT\t08\t0\n",
      "AGAT\t09\t2\n",
      "AGAT\t10\t98\n",
      "AAAC\t07\t4\n",
      "AAAC\t08\t93\n",
      "AAAC\t09\t90\n",
      "AAAC\t10\t93\n",
      "ACAT\t07\t100\n",
      "ACAT\t08\t95\n",
      "ACAT\t09\t99\n",
      "ACAT\t10\t89\n",
      "[[ 83. 100.  37.  22.]\n",
      " [ 26.  96.  99. 100.]\n",
      " [ 68.  80. 100.  99.]\n",
      " [ 99.  98. 100.  79.]\n",
      " [ 98.  25.  75.  60.]\n",
      " [100.  19.  94.  37.]\n",
      " [  2.   0.   2.  98.]\n",
      " [  4.  93.  90.  93.]\n",
      " [100.  95.  99.  89.]]\n",
      "CPU times: user 1min 1s, sys: 150 ms, total: 1min 1s\n",
      "Wall time: 1min 3s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
