{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write commands to run joint method validation \n",
    "# Without genotyping errors: Lookup table eurodem_prior2\n",
    "# With genotyping errors: Lookup table eudrodem_prior2_errors\n",
    "\n",
    "def validation():\n",
    "    \n",
    "    #means = [0.00001, 0.00005, 0.0001, 0.0005, 0.001, 0.005, 0.01] \n",
    "    #means = [10**-7,10**-6,10**-5,10**-4,10**-3,10**-2,10**-1]\n",
    "    means = [10**-6,10**-5,10**-4,10**-3,10**-2,10**-1]\n",
    "    pers = [2,3,4]\n",
    "    opt_dic = {}\n",
    "    opt_dic[2] = [11,20]\n",
    "    opt_dic[3] = [5,13]\n",
    "    opt_dic[4] = [7,10]\n",
    "    folder = 'ks_test_1000_genotyping_errors'\n",
    "    a_list = [0.2,0.3,0.6] \n",
    "    \n",
    "    num_loci = 1000\n",
    "    num_sims = 2000\n",
    "    num_trials = 20\n",
    "    \n",
    "    genotyping_errors = 1\n",
    "    \n",
    "    ks_test_threshold = 0.05\n",
    "    \n",
    "    num_files = 8\n",
    "\n",
    "    outFile = '/gymreklab-tscc/bonnieh/joint_method/commands_validation_top_x_%s.txt'%(folder)\n",
    "\n",
    "    results = open(outFile, \"w\")\n",
    "    \n",
    "    i = 0\n",
    "    for mean in means:\n",
    "        for a in a_list:\n",
    "            b = mean/a\n",
    "            for per in pers:\n",
    "                for opt in opt_dic[per]:\n",
    "                    i = i + 1\n",
    "                    results.write('time python \\/projects\\/ps-gymreklab\\/bonnieh\\/joint_method\\/joint_method_validation.py --period %d --out-subfolder %s --a-val %.9f --validation-mean %.8f --opt-allele-to-validate %d --num-loci %d --top-x --genotyping-errors --ks-test'%(per, folder,  a,  mean, opt, num_loci))\n",
    "                    if i%num_files == 0:\n",
    "                        results.write('\\t%d\\t%d\\t%.2f\\t%.8f\\n'%(per, opt, a, mean))\n",
    "                    else:\n",
    "                        results.write(' ; ')\n",
    "    results.write('\\t%d\\t%d\\t%.2f\\t%.8f\\n'%(per, opt, a, mean))\n",
    "    results.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write commands to run joint method on real data for different motifs\n",
    "# One job for each mutation rate setting\n",
    "# To run heatmap, requires 20 jobs (20 core hours?)\n",
    "\n",
    "def feas_mut_rates():\n",
    "    inFile = '1kg_allele_freqs_filt_intergenic.txt'\n",
    "    \n",
    "    opt_dic = {}\n",
    "    \n",
    "    opt_dic[2] = [11,12,13,14,15,16,17,18,19,20]\n",
    "    opt_dic[3] = [5,6,7,8,9,10,11,12,13]\n",
    "    opt_dic[4] = [7,8,9,10]\n",
    "    \n",
    "    motif_dic = {}\n",
    "    motif_dic[2] = ['AC','AG','AT'] #30\n",
    "    motif_dic[3] = ['ACC','AGG','AGC','CCG','AAG','AAC','ATC','AAT'] #72\n",
    "    motif_dic[4] = ['AAGG','AAAT','ATCC','AATC','AAAC','AATG','ACAT','AGAT','AAAG'] #36\n",
    "    \n",
    "    folder_name = '1kg_bin_agnostic_071921'\n",
    "    superFolder = 'results_intergenic_top_x' \n",
    "    outFile = '/gymreklab-tscc/bonnieh/joint_method/commands_intergenic_top_x_%s.txt'%(folder_name)\n",
    "    \n",
    "    num_loci = 1000 \n",
    "    num_sims = 50000 \n",
    "\n",
    "    results = open(outFile, \"w\")\n",
    "    \n",
    "    mut_setting = 'all'\n",
    "    pers = [2,3,4] \n",
    "   \n",
    "    num_files = 10\n",
    "    count = 0\n",
    "    \n",
    "    for per in pers:\n",
    "        for motif in motif_dic[per]:\n",
    "            opt_alleles = opt_dic[per]\n",
    "            for opt in opt_alleles:\n",
    "                results.write('time python \\/projects\\/ps-gymreklab\\/bonnieh\\/joint_method\\/SISTR_joint_method.py --motif-format --period %d --sim-model all --out-subfolder %s --opt-allele-to-use %d --motif-to-use %s --file-to-use %s --top-x'%(per, folder_name,  opt, motif, inFile))\n",
    "                results.write(' ; ')\n",
    "                count = count + 1\n",
    "                if (count % num_files) == 0:\n",
    "                    results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, mut_setting, opt))\n",
    "    if (count % num_files) != 0:\n",
    "        results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, mut_setting, opt))\n",
    "                \n",
    "    results.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write commands to run joint method on real data for different motifs\n",
    "# One job for each mutation rate setting\n",
    "# To run heatmap, requires 20 jobs (20 core hours?)\n",
    "\n",
    "def subsample():\n",
    "    inFile = '1kg_allele_freqs_filt_intergenic.txt'\n",
    "    \n",
    "    opt_dic = {}\n",
    "    \n",
    "    opt_dic[2] = [11,12,13,14,15,16,17,18,19,20]\n",
    "    opt_dic[3] = [5,6,7,8,9,10,11,12,13]\n",
    "    opt_dic[4] = [7,8,9,10]\n",
    "    \n",
    "    motif_dic = {}\n",
    "    motif_dic[2] = ['AC','AG','AT'] #30\n",
    "    motif_dic[3] = ['ACC','AGG','AGC','CCG','AAG','AAC','ATC','AAT'] #72\n",
    "    motif_dic[4] = ['AAGG','AAAT','ATCC','AATC','AAAC','AATG','ACAT','AGAT','AAAG'] #36\n",
    "    \n",
    "    folder_name = '1kg_bin_agnostic_subsample_50_loci'\n",
    "    superFolder = 'results_intergenic_top_x' \n",
    "    outFile = '/gymreklab-tscc/bonnieh/joint_method/commands_intergenic_top_x_%s.txt'%(folder_name)\n",
    "    \n",
    "    num_loci = 50 \n",
    "    num_sims = 50000 \n",
    "\n",
    "    results = open(outFile, \"w\")\n",
    "    \n",
    "    mut_setting = 'all'\n",
    "    pers = [2,3,4] \n",
    "    \n",
    "    num_files = 10000\n",
    "    count = 0\n",
    "    \n",
    "    for per in pers:\n",
    "        \n",
    "        for motif in motif_dic[per]:\n",
    "            opt_alleles = opt_dic[per]\n",
    "            for opt in opt_alleles:\n",
    "                results.write('time python \\/projects\\/ps-gymreklab\\/bonnieh\\/joint_method\\/SISTR_joint_method.py --motif-format --period %d --sim-model all --out-subfolder %s --opt-allele-to-use %d --motif-to-use %s --file-to-use %s --num-loci %d --top-x'%(per, folder_name,  opt, motif, inFile, num_loci))\n",
    "                results.write(' ; ')\n",
    "                count = count + 1\n",
    "                if (count % num_files) == 0:\n",
    "                    results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, mut_setting, opt))\n",
    "    if (count % num_files) != 0:\n",
    "        results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, mut_setting, opt))\n",
    "                \n",
    "    results.close() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write commands to run joint method on real data for different motifs\n",
    "# One job for each mutation rate setting\n",
    "# To run heatmap, requires 20 jobs (20 core hours?)\n",
    "\n",
    "def functional_categories():\n",
    "    inFile = '1kg_allele_freqs.txt'\n",
    "    \n",
    "    categs = [7,8,9,10,11,12]\n",
    "    opt_dic = {}\n",
    "    \n",
    "    opt_dic[2] = [11,12,13,14,15,16,17,18,19,20]\n",
    "    opt_dic[3] = [5,6,7,8,9,10,11,12,13]\n",
    "    opt_dic[4] = [7,8,9,10]\n",
    "    \n",
    "    motif_dic = {}\n",
    "    motif_dic[2] = ['AC','AG','AT'] #30\n",
    "    motif_dic[3] = ['ACC','AGG','AGC','CCG','AAG','AAC','ATC','AAT'] #72\n",
    "    motif_dic[4] = ['AAGG','AAAT','ATCC','AATC','AAAC','AATG','ACAT','AGAT','AAAG'] #36\n",
    "    \n",
    "    folder_name = 'func_categ_analysis'\n",
    "    superFolder = 'results_top_x' \n",
    "    outFile = '/gymreklab-tscc/bonnieh/joint_method/commands_top_x_%s.txt'%(folder_name)\n",
    "    \n",
    "    num_loci = 1000 \n",
    "    num_sims = 50000 \n",
    "\n",
    "    results = open(outFile, \"w\")\n",
    "    \n",
    "    mut_setting = 'eurodem_prior2_trinuc_e_1kg_euro'\n",
    "    pers = [3] #2,3,4\n",
    "   \n",
    "    #mut_settings = [(2, 'eurodem_prior2_dinuc_a')]\n",
    "    num_files = 9\n",
    "    count = 0\n",
    "    \n",
    "    for per in pers:\n",
    "        for column in categs:\n",
    "            opt_alleles = opt_dic[per]\n",
    "            for opt in opt_alleles:\n",
    "                results.write('time python \\/projects\\/ps-gymreklab\\/bonnieh\\/joint_method\\/SISTR_joint_method.py --motif-format --period %d --sim-model %s --out-subfolder %s --opt-allele-to-use %d --motif-to-use %s --file-to-use %s --column %d --out-folder %s --top-x'%(per, mut_setting, folder_name,  opt, '1.0', inFile, column, superFolder))\n",
    "                results.write(' ; ')\n",
    "                count = count + 1\n",
    "                if (count % num_files) == 0:\n",
    "                    results.write('\\t%d\\t%d\\t%s\\t%d\\n'%(per, column, mut_setting, opt))\n",
    "    if (count % num_files) != 0:\n",
    "        results.write('\\t%d\\t%d\\t%s\\t%d\\n'%(per, column, mut_setting, opt))\n",
    "                \n",
    "    results.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def motif_analysis_diff_mu():\n",
    "    inFile = '1kg_allele_freqs.txt'\n",
    "    opt_dic = {}\n",
    "    \n",
    "    opt_dic[2] = [11,12,13,14,15,16,17,18,19,20]\n",
    "    opt_dic[3] = [5,6,7,8,9,10,11,12,13]\n",
    "    opt_dic[4] = [7,8,9,10]\n",
    "    \n",
    "    motif_dic = {}\n",
    "    motif_dic[2] = ['AC','AG','AT'] #30\n",
    "    motif_dic[3] = ['ACC','AGG','AGC','CCG','AAG','AAC','ATC','AAT'] #72\n",
    "    motif_dic[4] = ['AAGG','AAAT','ATCC','AATC','AAAC','AATG','ACAT','AGAT','AAAG'] #36\n",
    "    \n",
    "    folder_name = 'motif_analysis_diff_mu_071921'\n",
    "    superFolder = 'results_top_x'\n",
    "    outFile = '/gymreklab-tscc/bonnieh/joint_method/commands_top_x_%s.txt'%(folder_name)\n",
    "    \n",
    "    num_loci = 1000 \n",
    "    num_sims = 50000 \n",
    "\n",
    "    results = open(outFile, \"w\")\n",
    "    mut_setting = {}\n",
    "    mut_setting['AC'] = 'eurodem_prior2_dinuc_e_1kg_euro'\n",
    "    mut_setting['AG'] = 'eurodem_prior2_dinuc_e_1kg_euro'\n",
    "    mut_setting['AT'] = 'eurodem_prior2_dinuc_d_1kg_euro'\n",
    "    for mot in motif_dic[3]:\n",
    "        mut_setting[mot] = 'eurodem_prior2_trinuc_e_1kg_euro'\n",
    "    for mot in ['AAAG', 'AAGG', 'AGAT']:\n",
    "        mut_setting[mot] = 'eurodem_prior2_tetranuc_b_1kg_euro'\n",
    "    for mot in ['ACAT', 'AAAT']:\n",
    "        mut_setting[mot] = 'eurodem_prior2_tetranuc_c_1kg_euro'\n",
    "    for mot in ['AATC', 'AATG', 'ATCC', 'AAAC']:\n",
    "        mut_setting[mot] = 'eurodem_prior2_tetranuc_d_1kg_euro'\n",
    "\n",
    "    pers = [2,3,4] #2,3,4\n",
    "    \n",
    "   \n",
    "    #mut_settings = [(2, 'eurodem_prior2_dinuc_a')]\n",
    "    num_files = 9\n",
    "    count = 0\n",
    "    \n",
    "    for per in pers:\n",
    "        for motif in motif_dic[per]:\n",
    "            opt_alleles = opt_dic[per]\n",
    "            for opt in opt_alleles:\n",
    "                results.write('time python \\/projects\\/ps-gymreklab\\/bonnieh\\/joint_method\\/SISTR_joint_method.py --motif-format --period %d --sim-model %s --out-subfolder %s --opt-allele-to-use %d --motif-to-use %s --file-to-use %s --out-folder %s --top-x'%(per, mut_setting[motif], folder_name,  opt, motif, inFile, superFolder))\n",
    "                results.write(' ; ')\n",
    "                count = count + 1\n",
    "                if (count % num_files) == 0:\n",
    "                    results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, motif, opt))\n",
    "    if (count % num_files) != 0:\n",
    "        results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, motif, opt))\n",
    "                \n",
    "    results.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def motif_analysis_dinuc_diff_setting():\n",
    "    inFile = '1kg_allele_freqs.txt'\n",
    "    \n",
    "    opt_dic = {}\n",
    "    \n",
    "    opt_dic[2] = [11,12,13,14,15,16,17,18,19,20]\n",
    "    \n",
    "    motif_dic = {}\n",
    "    motif_dic[2] = ['AC','AG','AT'] #30\n",
    "    \n",
    "    folder_name = 'motif_analysis_same_mu'#'motif_analysis_diff_beta_071921' 'motif_analysis_diff_rho_071921'\n",
    "    superFolder = 'results_top_x' #intergenic_bin_agnostic_upd'#ks_test_upd'\n",
    "    outFile = '/gymreklab-tscc/bonnieh/joint_method/commands_top_x_%s.txt'%(folder_name)\n",
    "    \n",
    "    num_loci = 1000 \n",
    "    num_sims = 50000 \n",
    "\n",
    "    results = open(outFile, \"w\")\n",
    "    mut_setting = {}\n",
    "    mut_setting['AC'] = 'eurodem_prior2_dinuc_e_1kg_euro'#'eurodem_prior2_dinuc_high_rho_1kg_euro'#'eurodem_prior2_dinuc_high_beta_1kg_euro'\n",
    "    mut_setting['AG'] = 'eurodem_prior2_dinuc_e_1kg_euro'#'eurodem_prior2_dinuc_high_rho_1kg_euro'#'eurodem_prior2_dinuc_high_beta_1kg_euro'\n",
    "    mut_setting['AT'] = 'eurodem_prior2_dinuc_e_1kg_euro'\n",
    "\n",
    "    pers = [2] \n",
    "    \n",
    "   \n",
    "  \n",
    "    num_files = 9\n",
    "    count = 0\n",
    "    \n",
    "    for per in pers:\n",
    "        for motif in motif_dic[per]:\n",
    "            opt_alleles = opt_dic[per]\n",
    "            for opt in opt_alleles:\n",
    "                results.write('time python \\/projects\\/ps-gymreklab\\/bonnieh\\/joint_method\\/SISTR_joint_method.py --motif-format --period %d --sim-model %s --out-subfolder %s --opt-allele-to-use %d --motif-to-use %s --file-to-use %s --out-folder %s --top-x'%(per, mut_setting[motif], folder_name,  opt, motif, inFile, superFolder))\n",
    "                results.write(' ; ')\n",
    "                count = count + 1\n",
    "                if (count % num_files) == 0:\n",
    "                    results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, motif, opt))\n",
    "    if (count % num_files) != 0:\n",
    "        results.write('\\t%d\\t%s\\t%s\\t%d\\n'%(per, motif, motif, opt))\n",
    "                \n",
    "    results.close() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    #validation()\n",
    "    #feas_mut_rates()\n",
    "    #subsample()\n",
    "    #motif_analysis_diff_mu()\n",
    "    #functional_categories()\n",
    "    motif_analysis_dinuc_diff_setting()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.58 ms, sys: 83 Âµs, total: 1.66 ms\n",
      "Wall time: 15.9 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
